src/main/kotlin/com/cognifide/gradle/aem/environment/EnvironmentPlugin.kt

```diff
@@ -0,0 +1,54 @@
+package com.cognifide.gradle.aem.environment
+
+import com.cognifide.gradle.aem.common.AemExtension
+import com.cognifide.gradle.aem.common.AemPlugin
+import com.cognifide.gradle.aem.environment.tasks.*
+import com.cognifide.gradle.aem.instance.InstancePlugin
+import com.cognifide.gradle.aem.instance.tasks.Destroy
+import com.cognifide.gradle.aem.instance.tasks.Setup
+import org.gradle.api.Project
+
+/**
+ * Separate plugin which provides tasks for managing local development environment additional to AEM,
+ * like Dispatcher, Solr, Knot.X, etc.
+ * Most often should be applied only to one project in build.
+ */
+class EnvironmentPlugin : AemPlugin() {
+
+    override fun Project.configure() {
+        setupDependentPlugins()
+        setupTasks()
+    }
+
+    private fun Project.setupDependentPlugins() {
+        plugins.apply(InstancePlugin::class.java)
+    }
+
+    private fun Project.setupTasks() {
+        with(AemExtension.of(this).tasks) {
+            register<EnvHosts>(EnvHosts.NAME)
+            register<EnvUp>(EnvUp.NAME)
+            register<EnvDown>(EnvDown.NAME)
+            register<EnvDestroy>(EnvDestroy.NAME) {
+                dependsOn(EnvDown.NAME)
+                finalizedBy(Destroy.NAME)
+            }
+            register<EnvSetup>(EnvSetup.NAME) {
+                dependsOn(Setup.NAME)
+                finalizedBy(EnvUp.NAME)
```

Comment:
generally the question is what is better.

https://github.com/Cognifide/gradle-aem-plugin/blob/master/src/main/kotlin/com/cognifide/gradle/aem/pkg/PackagePlugin.kt#L53

mustRunAfter or finalizedBy ... confusing.

Reply:
MustRunAfter isn't intuitive for me. Not sure what it will do. Here I want to run Setup and then EnvUp which is explicit from this code.